### DISCUSSION TITLE
blockstate support for blockTypeBreakRestriction allow/deny lists

### DISCUSSION THREAD

**Discussion by HaphLife** [maruohon/tweakeroo#347] [Ideas]
Hopefully a fairly small suggestion. I really hope there is a simple way to implement this:

Allow blockstates to be included in setting the allow and deny lists for blockTypeBreakRestriction.

Example: adding minecraft:nether_wart[age=2], would either allow or deny breaking specifically nether warts _with_ an age of two.

This is the same syntax as used by minecraft, for example /setblock ~ ~ ~ minecraft:nether_wart[age=2].

And that's the whole suggestion. Mainly I see using this passively to help not break crops of the wrong age, but it could be used situationally for redstone components facing particular directions, or stairs, or whatever else where you want to break one component in a tiling, but not others of the same block type.

**Comment by maruohon** [comment]
I wonder if the config part should then be implemented via a more complex system where you can create and switch between different "block state lists". Because changing string configs manually and often for stuff like this would be really awkward. Plus the new malilib configs would have had just a dropdown menu to select the block from for this config, instead of using string configs at all anymore. But if that dropdown would be the entire block state... it would look really messy in the dropdown due to the long block state string representation for some blocks. And it would still be pretty awkward to configure if it was just a "single config set" that you need to change every time.

**Reply by HaphLife** [reply]
Hmm. Yeah, implementing gui specifically for that does sound complicated. 
Presets sound kinda cool, but I can't really think of a particular usecase where you would frequently re-use a denylist that you don't want on by default. Maybe that makes more sense with an allowlist? (Like breaking snowlayers when you don't want to break the grass below, or stone in a stone generator or something)

As for the dropdowns, perhaps including something like command's tab complete would be easier to implement?

**Reply by maruohon** [reply]
The new malilib already has a "Block List" config type (as well as Item List and various other list types), which on the config screens is represented by a dropdown widget with a list of all the blocks in the game.

But for block states that can't be used as-is, it would at least need a separate list type that has all the blockstates. And that in itself is simple to do, but I'm pretty sure it would look messy in a dropdown widget, making it hard to find what you are looking for, even with the string search function that it has.

So I think a dedicated screen for selecting how to match block states would be in order for this type of config. And actually Litematica will also need very similar block state configs for some of the planned extra functionality/configs, so I'll most likely need to implement such a GUI in any case in malilib.

**Comment by HaphLife** [comment]
Gosh darnit. I was looking up how to use blockstates in the blacklist, and all I found was my own post requesting the feature to be added from the last time I tried to do this (which I had forgotten about). xD
